File: new-skills/references/beneficiary/inquiry-template/java/client/BeneficiaryManagementClient.java
/**
 * Alipay.com Inc.
 * Copyright (c) 2004-2026 All Rights Reserved.
 */
package {basePackage}.wf.client;

import com.alibaba.fastjson.JSON;
import {basePackage}.wf.config.WfConfig;
import {basePackage}.wf.model.exception.WfErrorCode;
import {basePackage}.wf.model.exception.WfException;
import {basePackage}.wf.model.request.BindBeneficiaryRequest;
import {basePackage}.wf.model.request.EditBeneficiaryRequest;
import {basePackage}.wf.model.request.InquiryBeneficiaryListRequest;
import {basePackage}.wf.model.request.InquiryBeneficiaryTemplateRequest;
import {basePackage}.wf.model.request.RemoveBeneficiaryRequest;
import {basePackage}.wf.model.response.BindBeneficiaryResponse;
import {basePackage}.wf.model.response.EditBeneficiaryResponse;
import {basePackage}.wf.model.response.InquiryBeneficiaryListResponse;
import {basePackage}.wf.model.response.InquiryBeneficiaryTemplateResponse;
import {basePackage}.wf.model.response.RemoveBeneficiaryResponse;
import {basePackage}.wf.model.response.Result;
import {basePackage}.wf.signer.WfSigner;
import {basePackage}.wf.util.WfHttpClientUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * WorldFirst 收款人管理统一客户端。
 *
 * <p>包含 5 个接口方法：
 * <ul>
 *   <li>{@link #inquiryBeneficiaryTemplate} — 查询卡模版</li>
 *   <li>{@link #bindBeneficiary} — 绑定收款人</li>
 *   <li>{@link #removeBeneficiary} — 删除收款人</li>
 *   <li>{@link #editBeneficiary} — 编辑收款人昵称</li>
 *   <li>{@link #inquiryBeneficiaryList} — 查询收款人列表</li>
 * </ul>
 *
 * @author Qoder
 * @version BeneficiaryManagementClient.java, v 0.1 2026-03-26
 */
public class BeneficiaryManagementClient {

    private static final Logger LOGGER = LoggerFactory.getLogger(BeneficiaryManagementClient.class);

    // API paths
    private static final String PATH_INQUIRY_TEMPLATE = "/amsin/api/v1/business/account/inquiryBeneficiaryTemplate";
    private static final String PATH_BIND = "/amsin/api/v1/business/account/bindBeneficiary";
    private static final String PATH_REMOVE = "/amsin/api/v1/business/account/removeBeneficiary";
    private static final String PATH_EDIT = "/amsin/api/v1/business/account/editBeneficiary";
    private static final String PATH_LIST = "/amsin/api/v1/business/account/inquiryBeneficiaryList";

    private static final int MAX_COUNTRY_CODE_LENGTH = 2;
    private static final int MAX_REQUEST_ID_LENGTH = 64;
    private static final int MAX_PAGE_SIZE = 50;

    private final WfConfig config;
    private WfHttpClientUtil httpClientUtil;
    private WfSigner signer;

    public BeneficiaryManagementClient(WfConfig config) {
        this.config = config;
    }

    public void init() {
        this.httpClientUtil = new WfHttpClientUtil(config);
        this.signer = httpClientUtil.getSigner();
    }

    /**
     * 允许外部注入 {@link WfHttpClientUtil}。
     * 适用于测试场景需要 mock HTTP 客户端的情况。
     *
     * @param httpClientUtil HTTP 客户端工具
     */
    public void setHttpClientUtil(WfHttpClientUtil httpClientUtil) {
        this.httpClientUtil = httpClientUtil;
        if (httpClientUtil != null) {
            this.signer = httpClientUtil.getSigner();
        }
    }

    // -------------------------------------------------------------------------
    // 1. 查询卡模版
    // -------------------------------------------------------------------------

    /**
     * 查询卡模版信息。
     *
     * @param request 查询请求
     * @return 卡模版响应
     * @throws WfException 参数校验失败或业务失败时抛出
     */
    public InquiryBeneficiaryTemplateResponse inquiryBeneficiaryTemplate(InquiryBeneficiaryTemplateRequest request) {
        validateTemplateRequest(request);
        return executeRequest(request, PATH_INQUIRY_TEMPLATE, InquiryBeneficiaryTemplateResponse.class);
    }

    // -------------------------------------------------------------------------
    // 2. 绑定收款人
    // -------------------------------------------------------------------------

    /**
     * 绑定收款人。
     *
     * @param request 绑定请求
     * @return 绑定响应（含 beneficiaryToken）
     * @throws WfException 参数校验失败或业务失败时抛出
     */
    public BindBeneficiaryResponse bindBeneficiary(BindBeneficiaryRequest request) {
        validateBindRequest(request);
        return executeRequest(request, PATH_BIND, BindBeneficiaryResponse.class);
    }

    // -------------------------------------------------------------------------
    // 3. 删除收款人
    // -------------------------------------------------------------------------

    /**
     * 删除收款人。
     *
     * @param request 删除请求
     * @return 删除响应
     * @throws WfException 参数校验失败或业务失败时抛出
     */
    public RemoveBeneficiaryResponse removeBeneficiary(RemoveBeneficiaryRequest request) {
        validateRemoveRequest(request);
        return executeRequest(request, PATH_REMOVE, RemoveBeneficiaryResponse.class);
    }

    // -------------------------------------------------------------------------
    // 4. 编辑收款人
    // -------------------------------------------------------------------------

    /**
     * 编辑收款人昵称。
     *
     * @param request 编辑请求
     * @return 编辑响应
     * @throws WfException 参数校验失败或业务失败时抛出
     */
    public EditBeneficiaryResponse editBeneficiary(EditBeneficiaryRequest request) {
        validateEditRequest(request);
        return executeRequest(request, PATH_EDIT, EditBeneficiaryResponse.class);
    }

    // -------------------------------------------------------------------------
    // 5. 查询收款人列表
    // -------------------------------------------------------------------------

    /**
     * 分页查询收款人列表。
     *
     * @param request 查询请求
     * @return 收款人列表响应
     * @throws WfException 参数校验失败或业务失败时抛出
     */
    public InquiryBeneficiaryListResponse inquiryBeneficiaryList(InquiryBeneficiaryListRequest request) {
        validateListRequest(request);
        return executeRequest(request, PATH_LIST, InquiryBeneficiaryListResponse.class);
    }

    // -------------------------------------------------------------------------
    // Validation methods
    // -------------------------------------------------------------------------

    private void validateTemplateRequest(InquiryBeneficiaryTemplateRequest request) {
        if (request == null) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "Request must not be null");
        }
        String countryCode = request.getCountryCode();
        if (countryCode != null && countryCode.length() > MAX_COUNTRY_CODE_LENGTH) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "countryCode max length is " + MAX_COUNTRY_CODE_LENGTH);
        }
    }

    private void validateBindRequest(BindBeneficiaryRequest request) {
        if (request == null) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "Request must not be null");
        }
        validateRequestId(request.getBindBeneficiaryRequestId(), "bindBeneficiaryRequestId");
        if (request.getBeneficiaryType() == null || request.getBeneficiaryType().trim().isEmpty()) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "beneficiaryType is required");
        }
    }

    private void validateRemoveRequest(RemoveBeneficiaryRequest request) {
        if (request == null) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "Request must not be null");
        }
        validateRequestId(request.getRemoveBeneficiaryRequestId(), "removeBeneficiaryRequestId");
        if (request.getBeneficiaryToken() == null || request.getBeneficiaryToken().trim().isEmpty()) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "beneficiaryToken is required");
        }
    }

    private void validateEditRequest(EditBeneficiaryRequest request) {
        if (request == null) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "Request must not be null");
        }
        if (request.getBeneficiaryToken() == null || request.getBeneficiaryToken().trim().isEmpty()) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "beneficiaryToken is required");
        }
        if (request.getBeneficiaryNick() == null || request.getBeneficiaryNick().trim().isEmpty()) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "beneficiaryNick is required");
        }
    }

    private void validateListRequest(InquiryBeneficiaryListRequest request) {
        if (request == null) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "Request must not be null");
        }
        if (request.getPageSize() == null || request.getPageSize() <= 0) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "pageSize is required and must be positive");
        }
        if (request.getPageSize() > MAX_PAGE_SIZE) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "pageSize max is " + MAX_PAGE_SIZE);
        }
        if (request.getPageNumber() == null || request.getPageNumber() <= 0) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, "pageNumber is required and must be positive");
        }
    }

    private void validateRequestId(String requestId, String fieldName) {
        if (requestId == null || requestId.trim().isEmpty()) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, fieldName + " is required");
        }
        if (requestId.length() > MAX_REQUEST_ID_LENGTH) {
            throw new WfException(WfErrorCode.PARAM_ILLEGAL, fieldName + " max length is " + MAX_REQUEST_ID_LENGTH);
        }
    }

    // -------------------------------------------------------------------------
    // Execute request
    // -------------------------------------------------------------------------

    private <T> T executeRequest(Object request, String apiPath, Class<T> responseClass) {
        String requestBody = JSON.toJSONString(request);
        String url = config.getBaseUrl() + apiPath;

        LOGGER.info("BeneficiaryManagementClient calling {}, url={}", apiPath, url);

        String responseBody = httpClientUtil.sendPostRequest(url, apiPath, requestBody);

        return parseResponse(responseBody, responseClass);
    }

    private <T> T parseResponse(String responseBody, Class<T> responseClass) {
        T response;
        try {
            response = JSON.parseObject(responseBody, responseClass);
        } catch (Exception e) {
            LOGGER.error("Failed to parse response: {}", responseBody, e);
            throw new WfException(WfErrorCode.INVALID_RESPONSE_FORMAT,
                "Failed to parse response: " + e.getMessage(), e);
        }

        // 获取 result 字段
        Result result = null;
        try {
            java.lang.reflect.Method getResult = responseClass.getMethod("getResult");
            result = (Result) getResult.invoke(response);
        } catch (Exception e) {
            LOGGER.warn("Response class has no getResult method");
        }

        if (result == null) {
            throw new WfException(WfErrorCode.INVALID_RESPONSE_FORMAT, "Response result is null");
        }

        String resultStatus = result.getResultStatus();
        if ("S".equals(resultStatus)) {
            LOGGER.info("BeneficiaryManagementClient success, resultCode={}", result.getResultCode());
            return response;
        }

        WfErrorCode errorCode = WfErrorCode.fromCode(result.getResultCode());
        LOGGER.error("BeneficiaryManagementClient failed, resultStatus={}, resultCode={}, message={}",
            resultStatus, result.getResultCode(), result.getResultMessage());
        throw new WfException(errorCode,
            "API call failed: " + result.getResultCode() + " - " + result.getResultMessage());
    }
}

File: new-skills/references/beneficiary/golang/client/beneficiary_management_client.go
package client

import (
	"encoding/json"
	"fmt"
	"log"

	"{moduleName}/wf/config"
	"{moduleName}/wf/model/exception"
	"{moduleName}/wf/model/request"
	"{moduleName}/wf/model/response"
	"{moduleName}/wf/signer"
	"{moduleName}/wf/util"
)

const (
	// API paths for beneficiary management
	inquiryBeneficiaryTemplatePath = "/amsin/api/v1/business/account/inquiryBeneficiaryTemplate"
	bindBeneficiaryPath            = "/amsin/api/v1/business/account/bindBeneficiary"
	removeBeneficiaryPath          = "/amsin/api/v1/business/account/removeBeneficiary"
	editBeneficiaryPath            = "/amsin/api/v1/business/account/editBeneficiary"
	inquiryBeneficiaryListPath     = "/amsin/api/v1/business/account/inquiryBeneficiaryList"
)

// BeneficiaryManagementClient is a unified client for all beneficiary management APIs.
// It provides 5 interface methods:
//   1. InquiryBeneficiaryTemplate - Query card template (查询卡模版)
//   2. BindBeneficiary - Bind beneficiary (绑定收款人)
//   3. RemoveBeneficiary - Remove beneficiary (删除收款人)
//   4. EditBeneficiary - Edit beneficiary nickname (编辑收款人昵称)
//   5. InquiryBeneficiaryList - Query beneficiary list (查询收款人列表)
type BeneficiaryManagementClient struct {
	config         *config.WfConfig
	httpClientUtil *util.WfHttpClient
	signer         signer.Signer
}

// NewBeneficiaryManagementClient creates a new BeneficiaryManagementClient.
func NewBeneficiaryManagementClient(cfg *config.WfConfig, s signer.Signer) *BeneficiaryManagementClient {
	return &BeneficiaryManagementClient{
		config:         cfg,
		httpClientUtil: util.NewWfHttpClient(cfg, s),
		signer:         s,
	}
}

// -------------------------------------------------------------------------
// 1. InquiryBeneficiaryTemplate - Query card template
// -------------------------------------------------------------------------

// InquiryBeneficiaryTemplate queries the card template for a country/currency/beneficiary type.
// Use the returned field definitions to determine which fields are required when binding a beneficiary.
//
// Parameters:
//   - req: The inquiry request containing countryCode, currency, and/or beneficiaryType
//
// Returns:
//   - The template response containing cardTemplateData, localCardTemplateData, and crossBorderCardTemplateData
//   - Error if the request fails or response is invalid
func (c *BeneficiaryManagementClient) InquiryBeneficiaryTemplate(
	req *request.InquiryBeneficiaryTemplateRequest,
) (*response.InquiryBeneficiaryTemplateResponse, error) {
	if req == nil {
		return nil, exception.NewWfException(exception.ParamIllegal, "request must not be nil")
	}

	resp := &response.InquiryBeneficiaryTemplateResponse{}
	if err := c.executeRequest(req, inquiryBeneficiaryTemplatePath, resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// -------------------------------------------------------------------------
// 2. BindBeneficiary - Bind beneficiary
// -------------------------------------------------------------------------

// BindBeneficiary binds a new beneficiary to the WF account and returns a beneficiaryToken.
// The beneficiaryBankAccount fields should be filled according to the card template.
//
// Parameters:
//   - req: The bind request containing beneficiary details
//
// Returns:
//   - The bind response containing the beneficiaryToken
//   - Error if validation fails or the request fails
func (c *BeneficiaryManagementClient) BindBeneficiary(
	req *request.BindBeneficiaryRequest,
) (*response.BindBeneficiaryResponse, error) {
	if err := req.Validate(); err != nil {
		return nil, exception.NewWfException(exception.ParamIllegal, err.Error())
	}

	resp := &response.BindBeneficiaryResponse{}
	if err := c.executeRequest(req, bindBeneficiaryPath, resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// -------------------------------------------------------------------------
// 3. RemoveBeneficiary - Remove beneficiary
// -------------------------------------------------------------------------

// RemoveBeneficiary removes a bound beneficiary identified by beneficiaryToken.
//
// Parameters:
//   - req: The remove request containing the beneficiaryToken to remove
//
// Returns:
//   - The remove response confirming the deletion
//   - Error if validation fails or the request fails
func (c *BeneficiaryManagementClient) RemoveBeneficiary(
	req *request.RemoveBeneficiaryRequest,
) (*response.RemoveBeneficiaryResponse, error) {
	if err := req.Validate(); err != nil {
		return nil, exception.NewWfException(exception.ParamIllegal, err.Error())
	}

	resp := &response.RemoveBeneficiaryResponse{}
	if err := c.executeRequest(req, removeBeneficiaryPath, resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// -------------------------------------------------------------------------
// 4. EditBeneficiary - Edit beneficiary nickname
// -------------------------------------------------------------------------

// EditBeneficiary updates the nick name of an existing beneficiary.
//
// Parameters:
//   - req: The edit request containing beneficiaryToken and new beneficiaryNick
//
// Returns:
//   - The edit response confirming the update
//   - Error if validation fails or the request fails
func (c *BeneficiaryManagementClient) EditBeneficiary(
	req *request.EditBeneficiaryRequest,
) (*response.EditBeneficiaryResponse, error) {
	if err := req.Validate(); err != nil {
		return nil, exception.NewWfException(exception.ParamIllegal, err.Error())
	}

	resp := &response.EditBeneficiaryResponse{}
	if err := c.executeRequest(req, editBeneficiaryPath, resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// -------------------------------------------------------------------------
// 5. InquiryBeneficiaryList - Query beneficiary list
// -------------------------------------------------------------------------

// InquiryBeneficiaryList queries the list of bound beneficiaries with pagination.
// pageSize max 50, pageNumber starts from 1.
//
// Parameters:
//   - req: The list request containing pagination and filter parameters
//
// Returns:
//   - The list response containing beneficiaries and pagination info
//   - Error if validation fails or the request fails
func (c *BeneficiaryManagementClient) InquiryBeneficiaryList(
	req *request.InquiryBeneficiaryListRequest,
) (*response.InquiryBeneficiaryListResponse, error) {
	if err := req.Validate(); err != nil {
		return nil, exception.NewWfException(exception.ParamIllegal, err.Error())
	}

	resp := &response.InquiryBeneficiaryListResponse{}
	if err := c.executeRequest(req, inquiryBeneficiaryListPath, resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// -------------------------------------------------------------------------
// Private helper methods
// -------------------------------------------------------------------------

// executeRequest is a generic helper to execute HTTP requests and parse responses.
func (c *BeneficiaryManagementClient) executeRequest(req interface{}, apiPath string, resp interface{}) error {
	requestBody, err := json.Marshal(req)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	url := c.config.BaseURL + apiPath
	log.Printf("[BeneficiaryManagementClient] Request URL: %s", url)

	responseBody, err := c.httpClientUtil.SendPostRequest(url, apiPath, string(requestBody))
	if err != nil {
		return fmt.Errorf("http request failed: %w", err)
	}

	return parseResponse(responseBody, resp)
}

// parseResponse parses the HTTP response and handles result status.
func parseResponse(responseBody string, resp interface{}) error {
	if err := json.Unmarshal([]byte(responseBody), resp); err != nil {
		return exception.NewWfException(exception.InvalidResponseFormat,
			fmt.Sprintf("failed to parse response: %v", err))
	}

	// Use type assertion to get the Result field
	result := getResultFromResponse(resp)
	if result == nil {
		return exception.NewWfException(exception.InvalidResponseFormat, "response result is nil")
	}

	switch result.ResultStatus {
	case "S":
		log.Printf("[BeneficiaryManagementClient] Success, resultCode=%s", result.ResultCode)
		return nil
	case "F":
		return exception.NewWfException(exception.FromCode(result.ResultCode), result.ResultMessage)
	case "U":
		// Unknown status - may need retry
		return exception.NewWfException(exception.FromCode(result.ResultCode),
			fmt.Sprintf("unknown/retryable status: %s - %s", result.ResultStatus, result.ResultMessage))
	default:
		return exception.NewWfException(exception.InvalidResponseFormat,
			fmt.Sprintf("unknown resultStatus: %s", result.ResultStatus))
	}
}

// getResultFromResponse extracts the Result field from response using type assertion.
func getResultFromResponse(resp interface{}) *response.Result {
	// Type assertion for known response types
	switch r := resp.(type) {
	case *response.InquiryBeneficiaryTemplateResponse:
		return &r.Result
	case *response.BindBeneficiaryResponse:
		return &r.Result
	case *response.RemoveBeneficiaryResponse:
		return &r.Result
	case *response.EditBeneficiaryResponse:
		return &r.Result
	case *response.InquiryBeneficiaryListResponse:
		return &r.Result
	default:
		return nil
	}
}
